dali/pipeline/operators/arithmetic/arithmetic_meta.h

```diff
@@ -0,0 +1,266 @@
+REGISTER_TYPE_PROMOTION(int16_t,  double, double);
+REGISTER_TYPE_PROMOTION(uint16_t, double, double);
+REGISTER_TYPE_PROMOTION(int32_t,  double, double);
+REGISTER_TYPE_PROMOTION(uint32_t, double, double);
+REGISTER_TYPE_PROMOTION(int64_t,  double, double);
+REGISTER_TYPE_PROMOTION(uint64_t, double, double);
+
+REGISTER_TYPE_PROMOTION(uint8_t,  int8_t, int16_t);
+REGISTER_TYPE_PROMOTION(int16_t,  int8_t, int16_t);
+REGISTER_TYPE_PROMOTION(uint16_t, int8_t, int32_t);
+REGISTER_TYPE_PROMOTION(int32_t,  int8_t, int32_t);
+REGISTER_TYPE_PROMOTION(uint32_t, int8_t, int64_t);
+REGISTER_TYPE_PROMOTION(int64_t,  int8_t, int64_t);
+REGISTER_TYPE_PROMOTION(uint64_t, int8_t, int64_t); //should be int128_t
+
+REGISTER_TYPE_PROMOTION(int16_t,  uint8_t, int16_t);
+REGISTER_TYPE_PROMOTION(uint16_t, uint8_t, uint16_t);
+REGISTER_TYPE_PROMOTION(int32_t,  uint8_t, int32_t);
+REGISTER_TYPE_PROMOTION(uint32_t, uint8_t, uint32_t);
+REGISTER_TYPE_PROMOTION(int64_t,  uint8_t, int64_t);
+REGISTER_TYPE_PROMOTION(uint64_t, uint8_t, uint64_t);
+
+REGISTER_TYPE_PROMOTION(uint16_t, int16_t, int32_t);
+REGISTER_TYPE_PROMOTION(int32_t,  int16_t, int32_t);
+REGISTER_TYPE_PROMOTION(uint32_t, int16_t, int64_t);
+REGISTER_TYPE_PROMOTION(int64_t,  int16_t, int64_t);
+REGISTER_TYPE_PROMOTION(uint64_t, int16_t, int64_t);
+
+REGISTER_TYPE_PROMOTION(int32_t,  uint16_t, int32_t);
+REGISTER_TYPE_PROMOTION(uint32_t, uint16_t, uint32_t);
+REGISTER_TYPE_PROMOTION(int64_t,  uint16_t, int64_t);
+REGISTER_TYPE_PROMOTION(uint64_t, uint16_t, uint64_t);
+
+REGISTER_TYPE_PROMOTION(uint32_t, int32_t, int64_t);
+REGISTER_TYPE_PROMOTION(int64_t,  int32_t, int64_t);
+REGISTER_TYPE_PROMOTION(uint64_t, int32_t, int64_t);
+
+REGISTER_TYPE_PROMOTION(int64_t,  uint32_t, int64_t);
+REGISTER_TYPE_PROMOTION(uint64_t, uint32_t, uint64_t);
+
+REGISTER_TYPE_PROMOTION(uint64_t, int64_t, int64_t);
+
+template <typename L, typename R>
+using binary_result_t = typename binary_op_promotion<L, R>::type;
+
+// TODO(klecki): It's probably and overkill, but I wrote the table above first
+// To reduce type switching and compilation times, it would be nice to get a function that
+// does the runtime type promotions probably with DALI ids.
+inline DALIDataType TypePromotion(DALIDataType left, DALIDataType right) {
+  DALIDataType result = DALIDataType::DALI_NO_TYPE;
+  TYPE_SWITCH(left, type2id, Left_t,
+    (int8_t, uint8_t, int16_t, uint16_t, int32_t, uint32_t,
+        int64_t, uint64_t, float16, float, double),
+    (
+      TYPE_SWITCH(right, type2id, Right_t,
+        (int8_t, uint8_t, int16_t, uint16_t, int32_t, uint32_t,
+            int64_t, uint64_t, float16, float, double),
+        (
+          using Result_t = binary_result_t<Left_t, Right_t>;
+          result = TypeInfo::Create<Result_t>().id();
+        ),
+        (DALI_FAIL("No suitable type found");)
+      );
+    ),
+    (DALI_FAIL("No suitable type found");)
+  );
+  return result;
+}
+
+template <ArithmeticOp op>
+struct arithm_meta {
```

Comment:
Please add description, what is this struct for. Just to be clear, "Struct for holding artithmetic metadata" is obvious, there should be added, "What is the metadata about?"

Reply:
Done